From efadefb83b24cc06de877b6fac8c90644b68cd56 Mon Sep 17 00:00:00 2001
From: Marc Mutz <marc.mutz@qt.io>
Date: Thu, 16 Oct 2025 11:03:46 +0200
Subject: [PATCH 437/553] QChar: fix the signature of foldCase(char32_t,
 char32_t&)

The only caller, ucstricmp(), passes UTF-16 code units, as implied by
the implementation of foldCase(char32_t, char32_t&) (which decodes
surrogate pairs, which shouldn't exist in UTF-32, and, if they do, not
be decoded). Of course, the accumulator variables were char32_t, but
only because C++ enforces this, due to the foldCase() signature.

Fix by taking both arguments as char16_t instead of char32_t.

As a drive-by, don't store the first argument in a char32_t and then
narrow it for the QChar::*surrogate*() calls, use the argument
instead.

Found by Clang 21's -Wcharacter-conversion.

Amends the start of the public history.

Pick-to: 6.8 6.5
Change-Id: Ic08b695749d7f68353a4af8703eb6a86ba88d567
Reviewed-by: Thiago Macieira <thiago.macieira@intel.com>
Reviewed-by: Ahmad Samir <a.samirh78@gmail.com>
(cherry picked from commit e2745f98b6fbd4818beb8336ebc9c69a9757bc1f)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
---
 src/corelib/text/qchar.cpp   | 6 +++---
 src/corelib/text/qstring.cpp | 4 ++--
 2 files changed, 5 insertions(+), 5 deletions(-)

diff --git a/src/corelib/text/qchar.cpp b/src/corelib/text/qchar.cpp
index 63dd3afdb69..1e4111646a8 100644
--- a/src/corelib/text/qchar.cpp
+++ b/src/corelib/text/qchar.cpp
@@ -1686,11 +1686,11 @@ static inline char32_t foldCase(const char16_t *ch, const char16_t *start)
     return convertCase_helper(ucs4, QUnicodeTables::CaseFold);
 }
 
-static inline char32_t foldCase(char32_t ch, char32_t &last) noexcept
+static inline char32_t foldCase(char16_t ch, char16_t &last) noexcept
 {
     char32_t ucs4 = ch;
-    if (QChar::isLowSurrogate(ucs4) && QChar::isHighSurrogate(last))
-        ucs4 = QChar::surrogateToUcs4(last, ucs4);
+    if (QChar::isLowSurrogate(ch) && QChar::isHighSurrogate(last))
+        ucs4 = QChar::surrogateToUcs4(last, ch);
     last = ch;
     return convertCase_helper(ucs4, QUnicodeTables::CaseFold);
 }
diff --git a/src/corelib/text/qstring.cpp b/src/corelib/text/qstring.cpp
index eea66810b0b..615cf3e6752 100644
--- a/src/corelib/text/qstring.cpp
+++ b/src/corelib/text/qstring.cpp
@@ -1197,8 +1197,8 @@ Q_NEVER_INLINE static int ucstricmp(qsizetype alen, const char16_t *a, qsizetype
     if (a == b)
         return qt_lencmp(alen, blen);
 
-    char32_t alast = 0;
-    char32_t blast = 0;
+    char16_t alast = 0;
+    char16_t blast = 0;
     qsizetype l = qMin(alen, blen);
     qsizetype i;
     for (i = 0; i < l; ++i) {
-- 
2.51.2

